dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,720 @@
+        )
+    if not pipeline._py_graph_built:
+        pipeline._build_graph()
+    input_node_names = []
+    for op in pipeline._ops:
+        if isinstance(op._op, _ExternalSource):
+            input_node_names.append(op.name)
+    return input_node_names
+
+
+class DALIOutputSampleRef:
+    """
+    Reference for a single sample output bound to a pipeline run.
+    """
+
+    def __init__(self, proxy, pipe_run_ref, output_idx, sample_idx):
+        """
+        Args:
+            proxy (_DALIProxy): The proxy object used for communication or data handling.
+            pipe_run_ref (DALIPipelineRunRef): A reference to the pipeline run.
+            output_idx (int): The index of the output in the pipeline.
+            sample_idx (int): The index of the sample within the batch.
+        """
+        self.proxy = proxy
+        self.pipe_run_ref = pipe_run_ref
+        self.output_idx = output_idx
+        self.sample_idx = sample_idx
+
+    def __repr__(self):
+        return (
+            f"DALIOutputSampleRef({self.pipe_run_ref}, "
+            + f"output_idx={self.output_idx}, sample_idx={self.sample_idx})"
+        )
+
+
+class DALIPipelineRunRef:
+    """
+    Reference for a DALI pipeline run iteration.
+    """
+
+    def __init__(self, batch_id):
+        """
+        Args:
+            batch_id (tuple(int, int)): A tuple that uniquely identifies the batch. The first
+                element represent the worker, and the second the batch index for that worker
+        """
+        self.batch_id = batch_id
+        self.call_args = {}
+        self.is_scheduled = False
+        self.is_complete = False
+
+    def __repr__(self):
+        return (
+            f"DALIPipelineRunRef(batch_id={self.batch_id}, call_args={self.call_args} "
+            f"is_scheduled={self.is_scheduled}, is_complete={self.is_complete})"
+        )
+
+
+class DALIOutputBatchRef:
+    """
+    Reference for a batched output bound to a pipeline run.
+    """
+
+    def __init__(self, pipe_run_ref, output_idx):
+        """
+        Args:
+            pipe_run_ref (DALIPipelineRunRef): A reference to the pipeline run.
+            output_idx (int): The index of the output in the pipeline.
+        """
+        self.pipe_run_ref = pipe_run_ref
+        self.output_idx = output_idx
+
+    def __repr__(self):
+        return f"DALIOutputBatchRef(pipe_run_ref={self.pipe_run_ref}, output_idx={self.output_idx})"
+
+
+def _collate_dali_output_sample_ref_fn(samples, *, collate_fn_map=None):
+    """
+    Special collate function that schedules a DALI iteration for execution
+    """
+    assert len(samples) > 0
+    pipe_run_ref = samples[0].pipe_run_ref
+    output_idx = samples[0].output_idx
+    proxy = samples[0].proxy
+    for i, sample in enumerate(samples):
+        if (
+            sample.proxy != proxy
+            or sample.pipe_run_ref != pipe_run_ref
+            or sample.output_idx != output_idx
+            or sample.sample_idx != i
```

Comment:
This line should produce a different error message.

Reply:
Will fix